File: Dynamic Programming/climbing_stairs.cc
// Link: https://leetcode.com/problems/climbing-stairs/

class Solution
{
    public:
        int climbs(int pos, vector<int>& scores)
        {
            if(scores[pos] == -1)
            {
                int counter = 0, n = scores.size();
                
                if(pos+1 == n)
                {
                    counter += 1;
                }
                else 
                {
                    if(pos+1<n)
                    {
                        counter += climbs(pos+1, scores); 
                    }

                    if(pos+2<n)
                    {
                        counter += climbs(pos+2, scores); 
                    }
                }
                
                scores[pos] = counter;
            }
            return scores[pos];
        }
    
        int climbStairs(int n)
        {
            if (n == 1 || n == 2)
            {
                return n;
            }
            
            vector<int> scores(n,-1);
            climbs(0, scores);
            
            /*
            for(int i=0;i<scores.size();i++)
            {
                cout<<scores[i]<<"\n";
            }
            */
            
            return scores[0]+scores[1];
        }
}; 

File: Dynamic Programming/climbing_stairs.py
# Link: https://leetcode.com/problems/climbing-stairs/

# Approach: The number of ways to reach current step depends on sum of number of ways to reach previous two steps.

class Solution(object):
    def climbStairs(self, n):
        ar = [0]*n
        
        if n == 0:
            return 0
        elif n == 1:
            ar[0] = 1
        else:
            ar[0], ar[1] = 1, 1
        
        for i in range(1, n):
            if i-1 >= 0:
                ar[i] += ar[i-1]
            
            if i-2 >= 0:
                ar[i] += ar[i-2]
                
        return ar[-1]

'''
# Other Solution
class Solution(object):
    def climbStairs(self, n):
        ar = [0]*n
        
        if n == 0:
            return 0
        elif n == 1:
            ar[0] = 1
        else:
            ar[0], ar[1] = 1, 2
        
        for i in range(2, n):
            ar[i] += ar[i-1] + ar[i-2]
                
        return ar[-1]
'''